### PR TITLE
Refactor api

### PR DESCRIPTION
**Pull request by ahesselgrave**
All handlers work (status not implemented yet)

*Known issues*
- StaticHandler implementation is a little janky, but works for now. Waiting on proper Init functionality with the NginxConfig blocks

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,7 +7,7 @@
 *.a
 
 # Ignore executables
-echo_server
+webserver
```
Comment (jobu95) [inline .gitignore:10]: SCOPE CREEP. jk

File: src/api/request_handler.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      request_handler_builders = new std::map<std::string, RequestHandler* (*)(void)>;
+    }
+    (*request_handler_builders)[type] = RequestHandlerRegisterer::Create;
+  }
+  static RequestHandler* Create() {
+    return new T;
+  }
+};
+
+#define REGISTER_REQUEST_HANDLER(ClassName) \
```
Comment (jobu95) [inline src/api/request_handler.h]: I understand this is the professor's code, but we should add some documentation briefly outlining how this macro works for future generations

File: src/api/response.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        HTTP_302_FOUND = 302,
+        HTTP_404_NOT_FOUND = 404
+    };
+    
+    // Simple accesors
+    std::string version() { return version_; }
+    ResponseCode status() { return status_; }
+    std::string status_message() { return status_message_; }
+    Headers headers() { return headers_; }
+    std::string body() { return body_; }
```
Comment (jobu95) [inline src/api/response.h:37]: Since the API kind of forced uppercase-starting camel cased methods on us, should we make every method the same way?

File: src/session.cc [hunk 4]
```diff
@@ -25,24 +25,21 @@ void Session::do_read()
...
+  std::string uri = req->uri();
   boost::system::error_code ec;
 
   // determine whether we have an echo handler or a response handler
 
-  RequestHandler *handler = NULL;
+  RequestHandler *handler = nullptr;
 
-  std::pair<const std::string, std::string> const *longestPrefixMapping = NULL;
+  std::pair<const std::string, std::string> const *longestPrefixMapping = nullptr;
```
Comment (jobu95) [inline src/session.cc:38]: too much C, thanks for fixing

**Comment by ahesselgrave** [comment]
lol I forgot that I had to update the config parser. I feel fuckin stupid but I forgot to put an include guard on it...will update that real quick

File: src/api/response.h (hunk 3 above)
Comment (ahesselgrave) [reply src/api/response.h:37]: for some reason the accessors were lower case but mutators had that weird camel case. Those functions came straight from the API so I just let them be.

File: src/api/request_handler.h (hunk 2 above)
Comment (ahesselgrave) [reply src/api/request_handler.h]: I'll put it in README.md then

Comment (ahesselgrave) [reply src/api/request_handler.h]: or actually, probably makes more sense in this header then.

File: .gitignore (hunk 1 above)
Comment (ahesselgrave) [reply .gitignore:10]: :rofl:

File: src/api/response.h (hunk 3 above)
Comment (jobu95) [reply src/api/response.h:37]: huh. bizarre. ok I guess I'll use that convention
